### ISSUE TITLE
Does the order of network interfaces have any relevance?

### ISSUE THREAD

**Issue by akarasulu** [subutai-io/snap#211]
I was told by Dima and Tina that network interfaces especially in virtual machines had to be in some order. Namely the bridged interface needed to be the first one. I think this was due to that interface name hardcoded roaming script. 

Does this still apply because this will change everything in my Vagrant machines. I'm doing stupid things now to make sure the first interface is the bridged one and Vagrant hates that. It might be the source of a lot of problems. If I can let this go it makes the Vagrant thing extremely trivial.

**Comment by Dilshat** [comment]
Yes there is a dependency of the roaming script on the order of interfaces. The bridged one must be the first among the 3 interfaces created by vagrant. No other way

**Comment by Dilshat** [comment]
See my comments https://github.com/subutai-io/snap/issues/176

**Comment by Dilshat** [comment]
There is another option:
Since it is hard to determine which interface is bridged from inside VM but easy from outside on host machine,
Vagrantfile should set some environment variable inside the guest or create some file in a specific location on the guest filesystem that contains the name of the bridged interface.
This way the roaming script can use this name in its operation and fallback to its current implementation if no name is set.

**Comment by akarasulu** [comment]
Yes this is one approach, we can manage the mechanism on vagrant boxes and images we produce but cannot handle situation where users install manually into a guest they prepare. We can update the documentation for this parameter.

This will not handle situations where users install the snap on existing non-hypervisor systems. But then again it's not intended to be in effect for non-virtual configurations so that's ok.  

Regardless I think we need an additional approach to compliment this one that you suggest. I think we can detect the hypervisor using dmidecode which is on most Linux distributions. Check this out here with [`dmidecode -s system-product-name`](https://unix.stackexchange.com/questions/89714/easy-way-to-determine-virtualization-technology). Knowing the hypervisor being used it's very easy to know the interface naming used. We can make some good guesses if this parameter is not set.

**Comment by akarasulu** [comment]
Take a look [here at this nasty way](https://github.com/subutai-io/packer/blob/master/http/fix-vagrant) I correct Vagrant's changes to boxes. If you make the first interface bridged vagrant will try to change that internally by mucking with your network configuration. I register for notifications to see when this happens and reverse Vagrant's changes. This is what I hope to get rid of before handing things off to sysnet with these packer images.

**Comment by Dilshat** [comment]
For now we need to pass the name of bridged interface or at least its order from Vagrantfile to inside of VM.
The detection of which hypervisor we are in will help just optimize the detection of interface names.
It does not help detect which on of them is bridged. So waiting from you for this parameter

**Comment by akarasulu** [comment]
I'm giving this Vagrant stuff to @jadilet ... please discuss with him @Dilshat - he'll ping to you too I think.

**Comment by Dilshat** [comment]
ok

**Comment by Dilshat** [comment]
As a quick solution we can rely on the fact that Vagrant registers NAT interface as the first one since it uses it for port forwarding and enabling `vagrant ssh` command.
So next one created in Vagrantfile should be registered as second and so on.
We can have a convention of registering a bridged network in the second place.

**Comment by akarasulu** [comment]
Perhaps we should just set a parameter that this was a box created by Vagrant. This is already done. The box creation time is put into a file in the `subutai` user's home directory. Knowing this we can appropriately manage interfaces in the roaming script. 

So are you recommending we abandon this ugly hack that force Vagrant to use the first bridged interface? I'm personally afraid that users will want to change their vm networking and it will always keep reverting without them knowing what is happening. That's why this hack is a bad approach IMHO.

**Comment by Dilshat** [comment]
By default Vagrantfile can put a file containing name of bridged interface in *ubuntu* user's home directory, say */home/ubuntu/bridged.iface.name* containing name *eth1* (hard assumption). This way user can change vagrantfile or directly the bridged.iface.name as needed.
If we use the same convention for the cloud peers then we can keep changes minimal.

**Comment by akarasulu** [comment]
This sounds good. Let's make sure @jadilet knows about this. Thanks D.

**Comment by Dilshat** [comment]
Additional question @akarasulu :+1: why do we need host-only network at all?

**Comment by akarasulu** [comment]
Heh, you tell me. I don't think we need it. What was it's purpose? NAT should be enough with the port forwarding. But those old ports are not even being used by tray anymore to manage the peer I think. So there's probably no more reason for host only. You should know better than me @Dilshat, worries that you're asking me :-).

**Comment by Dilshat** [comment]
Ok lets remove it then. This will simplify the roaming script.
